src/cvrp/ecvrp.py

```diff
@@ -80,58 +90,348 @@ def get_roads(self) -> tuple[tuple[int, ...], ...]:
+        else:
+            index_min = index2
+            index_max = index1
+
+        begining = road[0:index_min]
+        middle = road[index_min:(index_max+1)]
+        middle.reverse()
+        ending = road[(index_max+1):size]
+
+        reversed_road = begining + middle + ending
+
+        return reversed_road
+
+    def __two_opt(self, road: list[int]) -> list[int]:
+        """
+        2-opt variance algorithm.
+
+        Algorithm which find the best mutation for this road by calculating
+        the distance for each mutation of road. The shortest distance will
+        determine the best muatation road.
+        """
+        best_distance = self._compute_road_fitness(road)
+        best_road = road
+
+        road.remove(0)  # remove the first depot
+        road.remove(0)  # remove the last depot
+
+        for index_point1 in range(len(road)):
+            for index_point2 in range(len(road)):
+                if index_point1 != index_point2:
+                    tmp_road = self.__reversed_content(road, index_point1, index_point2)
+                    tmp_road.insert(0, 0)  # add the depot at the begining
+                    tmp_road.append(0)  # add the depot at the end
+                    distance = self._compute_road_fitness(tmp_road)
+                    if distance < best_distance:
+                        best_distance = distance
+                        best_road = tmp_road
+
+        return best_road
+
+    def delete_empty_road(self, solution: list[list[int]]) -> list[list[int]]:
+        """Delete empty road which are represented by [0, 0]."""
+        for road in solution:
+            if road == [0, 0]:
+                solution.remove(road)
+        return solution
+
     def mutate(self) -> None:
-        """ Mutate the current individual according to its internal rules.
-            This changement is done in place.
         """
+        Mutate the current individual according to its internal rules.
 
-    def crossover(self, other: TypeIndividual) -> list[TypeIndividual]:
-        """ Generate a list of children according to the rules of the individual.
-            The returned list might be empty or contains duplicated children.
-            This function might not return the same result with the same argument and
-            will probably not be comutative.
+        This changement is done in place.
         """
+        solution = self.get_roads()
+        solution = self.__tuple_to_list(solution)
+
+        choice = randrange(0, len(solution))
+        road = solution[choice]
+
+        road = self.__pull_off_chargers(road)
+
+        # algo 2-opt
+        mutant_road = self.__two_opt(road)
+
+        new_road = self._road_correction(mutant_road, self.__instance.get_ev_battery())
+        solution[choice] = new_road
+
+        solution = self.delete_empty_road(solution)
+
+        self._solution = self.__merge_roads(solution)
+
+    def crossover(self, other: "ECVRPSolution") -> list["ECVRPSolution"]:
```

Comment:
This method is quite long and has a lot of duplicated code in it. Could you consider refactoring it to avoid at least part of the duplication ?

Reply:
done